scrypto-schema/src/lib.rs

```diff
@@ -136,28 +152,44 @@ pub struct IndexedBlueprintSchema {
 
 impl From<BlueprintSchema> for IndexedBlueprintSchema {
     fn from(schema: BlueprintSchema) -> Self {
-        let mut module_offset = 0u8;
-
-        let tuple_module = if schema.substates.is_empty() {
-            None
-        } else {
-            let tuple_module = Some((module_offset, schema.substates));
-            module_offset += 1;
-            tuple_module
+        let mut partition_offset = 0u8;
+
+        let mut fields = None;
+        if !schema.fields.is_empty() {
+            fields = Some((PartitionOffset(partition_offset), schema.fields));
+            partition_offset += 1;
         };
 
-        let mut key_value_stores = Vec::new();
+        let mut collections = Vec::new();
+        for kv_schema in schema.kv_stores {
```

Comment:
I've got some pretty grave concerns about this strategy for auto-allocation of partition numbers when it comes to upgradability.

Adding an additional KVStore would change the partition numbers of the indices / sorted indices.

I think instead we probably need to explicitly claim indices for the relevant partitions in the schema itself (and validate that they make sense and won't overflow).

What do you think?

Reply:
Adding an additional KVStore wouldn't use the same interface and the idea would be for the backend to manage the mapping as new collections are added (the reason why offset is included in the collection vector).

I much prefer the user layer not needing to manage indices on it's own. All the user layer sees is a vector of collections (similar to fields).